### PR TITLE
Feature: Allow gs:// paths to be used in resources

### PR DESCRIPTION
**Pull request by cortadocodes**
### New Features
- [x] Allow `gs://` paths to be used in `Datafile`, `Dataset`, and `Manifest`
- [x] Allow `gs://` paths to be used in storage client 

### Minor fixes and improvements
- [x] Remove a line of duplicated code in `Datafile`

## Quality Checklist
- [x] New features are fully tested (No matter how much Coverage Karma you have)

### CODE REVIEW

File: octue/cloud/storage/client.py [hunk 1]
```diff
@@ -46,17 +47,21 @@ def create_bucket(self, name, location=None, allow_existing=False, timeout=_DEFA
 
         self.client.create_bucket(bucket_or_name=name, location=location, timeout=timeout)
 
-    def upload_file(self, local_path, bucket_name, path_in_bucket, metadata=None, timeout=_DEFAULT_TIMEOUT):
-        """Upload a local file to a Google Cloud bucket at gs://<bucket_name>/<path_in_bucket>.
+    def upload_file(
+        self, local_path, gs_path=None, bucket_name=None, path_in_bucket=None, metadata=None, timeout=_DEFAULT_TIMEOUT
```
Comment (thclark) [inline octue/cloud/storage/client.py:51]: I think gs_path should be called cloud_path for future API compatibility with other platforms

File: octue/resources/manifest.py [hunk 2]
```diff
@@ -45,18 +45,23 @@ def __init__(self, id=None, logger=None, path=None, datasets=None, keys=None, **
...
+    def from_cloud(cls, project_name, gs_path=None, bucket_name=None, path_to_manifest_file=None):
+        """Instantiate a Manifest from Google Cloud storage. Either (`bucket_name` and `path_to_manifest_file`) or
+        `gs_path` must be provided.
 
         :param str project_name:
-        :param str bucket_name:
-        :param str path_to_manifest_file:
+        :param str|None gs_path:
+        :param str|None bucket_name:
+        :param str|None path_to_manifest_file:
```
Comment (thclark) [inline octue/resources/manifest.py:55]: document params on public methods

File: octue/utils/encoders.py [hunk 3]
```diff
@@ -10,5 +12,8 @@ def default(self, obj):
         if hasattr(obj, "serialise"):
             return obj.serialise()
 
+        if isinstance(obj, datetime.datetime):
+            return str(obj)
+
```
Comment (thclark) [inline octue/utils/encoders.py:17]: this is the default method of serializing any object. Shouldn't be required?

**Review by thclark — CHANGES_REQUESTED** [review]

File: octue/utils/encoders.py (hunk 3 above)
Comment (cortadocodes) [reply octue/utils/encoders.py:17]: It wasn't working for some reason - it complains that `datetime`s aren't serialisable.

Comment (cortadocodes) [reply octue/utils/encoders.py:17]: I get this if I remove it:
```
Error
Traceback (most recent call last):
  File "/Users/Marcus1/repos/octue/octue-sdk-python/tests/resources/test_dataset.py", line 385, in test_to_cloud
    dataset.to_cloud(project_name, **location_parameters)
  File "/Users/Marcus1/repos/octue/octue-sdk-python/octue/resources/dataset.py", line 119, in to_cloud
    serialised_dataset = self.serialise()
  File "/Users/Marcus1/repos/octue/octue-sdk-python/octue/mixins/serialisable.py", line 93, in serialise
    string = json.dumps(self_as_primitive, cls=OctueJSONEncoder, sort_keys=True, indent=4, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/__init__.py", line 234, in dumps
    return cls(
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/encoder.py", line 201, in encode
    chunks = list(chunks)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/encoder.py", line 431, in _iterencode
    yield from _iterencode_dict(o, _current_indent_level)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/encoder.py", line 405, in _iterencode_dict
    yield from chunks
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/encoder.py", line 325, in _iterencode_list
    yield from chunks
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/encoder.py", line 438, in _iterencode
    o = _default(o)
  File "/Users/Marcus1/repos/octue/octue-sdk-python/octue/utils/encoders.py", line 13, in default
    return obj.serialise()
  File "/Users/Marcus1/repos/octue/octue-sdk-python/octue/mixins/serialisable.py", line 93, in serialise
    string = json.dumps(self_as_primitive, cls=OctueJSONEncoder, sort_keys=True, indent=4, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/__init__.py", line 234, in dumps
    return cls(
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/encoder.py", line 201, in encode
    chunks = list(chunks)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/encoder.py", line 431, in _iterencode
    yield from _iterencode_dict(o, _current_indent_level)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/encoder.py", line 405, in _iterencode_dict
    yield from chunks
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/encoder.py", line 405, in _iterencode_dict
    yield from chunks
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/encoder.py", line 438, in _iterencode
    o = _default(o)
  File "/Users/Marcus1/repos/octue/octue-sdk-python/octue/utils/encoders.py", line 19, in default
    return TwinedEncoder.default(self, obj)
  File "/Users/Marcus1/repos/octue/octue-sdk-python/venv/lib/python3.8/site-packages/twined/utils/encoders.py", line 32, in default
    return json.JSONEncoder.default(self, obj)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type datetime is not JSON serializable
```

Comment (thclark) [reply octue/utils/encoders.py:17]: Ok. So, instead of str I think we should be more careful and format it as RFC3339.

Comment (cortadocodes) [reply octue/utils/encoders.py:17]: Good point. I assume we want timezone information in there? How do we get timezone information if users don't provide it?

Comment (cortadocodes) [reply octue/utils/encoders.py:17]: Maybe we should just serialise it as is but in the `isoformat`?

Comment (thclark) [reply octue/utils/encoders.py:17]: > Maybe we should just serialise it as is but in the `isoformat`?

RFC3339 is a more tightly defined isoformat

Comment (thclark) [reply octue/utils/encoders.py]: to resolve this: we decided to isoformat() for simplicity, and let the user worry about timezone clusterfucks.

**Pull request merged**